aspen/request_processor/dispatcher.py

```diff
@@ -297,3 +355,198 @@ def _match_index(indices, indir):
         if os.path.isfile(index):
             return index
     return None
+
+
+class UserlandDispatcher(Dispatcher):
+    """This is Aspen's new dispatcher, optimized for production use.
+
+    This implementation builds a complete dispatch tree when it is first created.
+    That allows it to route requests efficiently, with fewer computations and
+    memory allocations than the old dispatcher, and without making any system
+    call, thus avoiding FFI and context switching costs as well.
+    """
+
+    DIR_WILDCARD = Constant('DIR_WILDCARD')
+    LEAF_WILDCARDS = Constant('LEAF_WILDCARDS')
+
+    collision_handler = staticmethod(legacy_collision_handler)
+    file_skipper = staticmethod(skip_hidden_files)
+
+    def build_dispatch_tree(self):
```

Comment:
This and `dispatch` have high cyclomatic complexity. Decomposing into multiple functions would make it easier to follow, imo, but I can live with this since I know your style differs and the tests pass. :)

Reply:
I think the functions mostly need more comments, although now that you mention it I do see ~10 lines that could be moved out of `build_dispatch_tree` and into a separate method.